src/security_utilities_rust/src/microsoft_security_utilities_core/cross_company_correlating_id.rs

```diff
@@ -2,34 +2,55 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 use base64::{engine::general_purpose::STANDARD, Engine as _};
-use sha2::{Sha256, Digest};
-use std::fmt;
+use sha2::{Digest, Sha256};
 use std::cell::RefCell;
 
 thread_local! {
     static THREAD_LOCAL_SHA256: RefCell<Sha256> = RefCell::new(Sha256::new());
 }
 
 pub fn generate_cross_company_correlating_id(text: &str) -> String {
-    let hash = generate_sha256_hash(text);
-
-    let hash = format!("CrossMicrosoftCorrelatingId:{}", hash);
+    let mut buffer = [0u8; 64];
+    let hash = generate_sha256_hash(text, &mut buffer);
 
     let checksum = THREAD_LOCAL_SHA256.with(|sha| {
-        sha.borrow_mut().update(hash.as_bytes());
-        sha.borrow_mut().finalize_reset()
+        let mut sha = sha.borrow_mut();
+
+        sha.update("CrossMicrosoftCorrelatingId:");
+        sha.update(hash);
+        sha.finalize_reset()
     });
 
     let to_encode = &checksum[0..15];
     STANDARD.encode(to_encode)
 }
 
-pub fn generate_sha256_hash(text: &str) -> String {
-
+fn generate_sha256_hash<'a>(text: &str, buffer: &'a mut [u8; 64]) -> &'a [u8; 64] {
     let result = THREAD_LOCAL_SHA256.with(|sha| {
         sha.borrow_mut().update(text.as_bytes());
         sha.borrow_mut().finalize_reset()
     });
 
-    fmt::format(format_args!("{:X}", result))
-}
\ No newline at end of file
+    let data: [u8; 32] = result.into();
+
+    for (idx, byte) in data.into_iter().enumerate() {
```

Comment:
@jdraaijer-microsoft I don't entirely follow the algorithm here. Since the tests pass, I have faith in the logic. But a question: generating cross company ID should be relatively infrequent. Are the gains with this optimization high enough to trade off with respect to the readability?

Reply:
It saves one allocation (~7% perf for few-block hashes). This is just a manual implementation of UpperHex formatting for the array.

I can remove it or add a clarifying comment

Reply:
I've added a comment explaining what this does. IMO it's worth it since the encoding is so simple, but I can agree that it is a minor nuisance to understand